Library/include/CSP/Multiplayer/EventBus.h

```diff
@@ -0,0 +1,110 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "CSP/Common/String.h"
+#include "CSP/Multiplayer/MultiPlayerConnection.h"
+#include "Multiplayer/EventSerialisation.h"
+
+#include <map>
+
+namespace csp::systems
+{
+
+class SpaceSystem;
+class SystemsManager;
+class SystemBase;
+
+} // namespace csp::systems
+
+namespace csp::memory
+{
+
+CSP_START_IGNORE
+template <typename T> void Delete(T* Ptr);
+CSP_END_IGNORE
+
+} // namespace csp::memory
+
+
+/// @brief Namespace that encompasses everything in the multiplayer system
+namespace csp::multiplayer
+{
+
+class ReplicatedValue;
+
+/// @ingroup Multiplayer
+/// @brief Handling of all network events.
+class CSP_API EventBus
+{
+public:
+	/** @cond DO_NOT_DOCUMENT */
+	friend class csp::systems::SpaceSystem;
+	friend class csp::systems::SystemsManager;
+	friend class MultiplayerConnection;
+	friend void csp::memory::Delete<EventBus>(EventBus* Ptr);
+	/** @endcond */
+
+	// The callback used to register to listen to events.
+	typedef std::function<void(bool, const csp::common::Array<ReplicatedValue>&)> ParameterisedCallbackHandler;
+
+	/// @brief Registers a system to listen for the named event, where the system can define its
+	/// @brief own callback and deserialiser.
+	/// @param EventName csp::common::String : The identifying name for the event to listen for.
+	/// @param System csp::systems::SystemBase* : A pointer to the system which wants to register for the event.
+	CSP_NO_EXPORT void ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System);
+
+	/// @brief Registers a callback to listen for the named event
+	/// @param EventName csp::common::String : The identifying name for the event to listen for.
+	/// @param Callback ParameterisedCallbackHandler : A callback to register for the event which contains the parameter payload data.
+	void ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback);
+
+	/// @brief Stops the event bus from listening for a particular event, for any system or callbacks
+	/// @brief that were registered.
+	/// @param EventName csp::common::String : The identifying name for the event to stop listening for.
+	void StopAllListenEvent(const csp::common::String& EventName);
+
+	/// @brief Stops the event bus from listening for a particular event, for the specified system.
+	/// @param EventName csp::common::String : The identifying name for the event to stop listening for.
+	/// @param System csp::systems::SystemBase : The system that will stop being registered.
+	void StopSystemListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System);
```

Comment:
The "ListenEvent" both have the same naming convention. Is there a reason the stop listening have different names?

Reply:
the reasoning was the following: there is a function that deregisters everything, called StopAllListenEvent. This function has a different name following Adam's comment that it should be crystal clear that it deletes everything (including all callbacks if several are defined). From this, it seemed a little odd to to me have 2 overloaded functions, one for a system, one for a callback. you are right that it is not following ListenEvent, and it is best to do the same for both. I will only keep StopAllListenEvent, and rename the other 2 to StopListenEvent

Reply:
Sounds good!